### PR TITLE
Fix when SimpleArithmetic is used in Pagination order by

### PR DESCRIPTION
**Pull request by Seb33300**
This fix Pagination limit subquery when a `SimpleArithmetic` is used in the `ORDER BY`.

### CODE REVIEW

**Comment by lcobucci** [comment]
@Seb33300 would please also add some functional tests that cover that use case?

You can find examples on https://github.com/doctrine/doctrine2/tree/388afb46d0cb3ed0c51332e8df0de9e942c2690b/tests/Doctrine/Tests/ORM/Functional/Ticket

**Comment by Seb33300** [comment]
@lcobucci done

**Comment by Seb33300** [comment]
It looks like PostgreSQL require the full ORDER BY expression in the SELECT to work...

> SQLSTATE[42P10]: Invalid column reference: 7 ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryWalker.php [hunk 1]
```diff
@@ -157,20 +159,20 @@ private function validate(SelectStatement $AST)
     /**
      * Retrieve either an IdentityFunction (IDENTITY(u.assoc)) or a state field (u.name).
      *
-     * @param \Doctrine\ORM\Query\AST\PathExpression $pathExpression
+     * @param mixed $expression
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryWalker.php]: This should not be `mixed`: let's try avoiding `mixed` overall

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH7805Test.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        $article = new CmsArticle();
+
+        $article->topic   = 'Test SimpleArithmetic ORDER BY';
+        $article->text    = 'This test fails on MySQL.';
+        $article->version = 1;
+
+        $this->_em->persist($article);
+        $this->_em->flush();
+
+        $query = $this->_em->createQuery('SELECT a FROM Doctrine\Tests\Models\CMS\CmsArticle a ORDER BY a.version + 0 ASC');
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/GH7805Test.php]: Does this now also work with custom DQL functions?

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH7805Test.php (hunk 2 above)
Comment (Seb33300) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH7805Test.php]: Originally, only `PathExpression` found in the `ORDER BY` part was added to the `SELECT`, everything else was skipped.

I just added another condition to also accept `SimpleArithmetic` (see line 102 upper in this file) so i don't think custom DQL functions will work unless they are considered as `PathExpression` or `SimpleArithmetic`.

Maybe we should consider that every `Node` found in the `ORDER BY` part could be added in the `SELECT` part and remove the condition on line 102?

File: lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryWalker.php (hunk 1 above)
Comment (Seb33300) [reply lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryWalker.php]: I can replace it by `Doctrine\ORM\Query\AST\Node`?

Comment (Ocramius) [reply lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryWalker.php]: That would be much better, thanks

File: tests/Doctrine/Tests/ORM/Functional/Ticket/GH7805Test.php (hunk 2 above)
Comment (Ocramius) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH7805Test.php]: I'm not sure why the `Identity` function wrapper was added in first place, but if you could add an SQL generation code using something like `SQRT()`, that would be sufficient to check how we need to modify the code.

Comment (Seb33300) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/GH7805Test.php]: Done

**Comment by Seb33300** [comment]
Can someone please remove the `Missing Tests` label?

File: tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php [hunk 3]
```diff
@@ -47,7 +47,21 @@ public function testLimitSubqueryWithSortFunction() : void
         $limitQuery->setHint(Query::HINT_CUSTOM_TREE_WALKERS, [LimitSubqueryWalker::class]);
 
         self::assertSame(
-            'SELECT DISTINCT m0_.id AS id_0 FROM MyBlogPost m0_ INNER JOIN Category c1_ ON m0_.category_id = c1_.id GROUP BY m0_.id ORDER BY COUNT(c1_.id) ASC',
+            'SELECT DISTINCT m0_.id AS id_0, COUNT(c1_.id) AS sclr_1 FROM MyBlogPost m0_ INNER JOIN Category c1_ ON m0_.category_id = c1_.id GROUP BY m0_.id ORDER BY COUNT(c1_.id) ASC',
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php]: This line is getting long, how about adding some linebreaks?
```suggestion
            <<<SQL
      SELECT DISTINCT m0_.id AS id_0, COUNT(c1_.id) AS sclr_1 
         FROM MyBlogPost m0_ 
INNER JOIN Category c1_ ON m0_.category_id = c1_.id 
 GROUP BY m0_.id 
 ORDER BY COUNT(c1_.id) ASC
 SQL
            ,
```

File: tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php [hunk 4]
```diff
@@ -47,7 +47,21 @@ public function testLimitSubqueryWithSortFunction() : void
...
+            $limitQuery->getSQL()
+        );
+    }
+
+    public function testLimitSubqueryWithSortSimpleArithmetic() : void
+    {
+        $dql   = 'SELECT p FROM Doctrine\Tests\ORM\Tools\Pagination\MyBlogPost p JOIN p.category c GROUP BY p.id ORDER BY p.title + 0';
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php]:
```suggestion
        $dql   = <<<'DQL'
     SELECT p 
        FROM Doctrine\Tests\ORM\Tools\Pagination\MyBlogPost p 
          JOIN p.category c 
GROUP BY p.id 
ORDER BY p.title + 0
DQL;
```

File: tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php [hunk 5]
```diff
@@ -47,7 +47,21 @@ public function testLimitSubqueryWithSortFunction() : void
...
+        $query = $this->entityManager->createQuery($dql);
+
+        $limitQuery = clone $query;
+        $limitQuery->setHint(Query::HINT_CUSTOM_TREE_WALKERS, [LimitSubqueryWalker::class]);
+
+        self::assertSame(
+            'SELECT DISTINCT m0_.id AS id_0, m0_.title + 0 AS sclr_1 FROM MyBlogPost m0_ INNER JOIN Category c1_ ON m0_.category_id = c1_.id GROUP BY m0_.id ORDER BY m0_.title + 0 ASC',
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php]:
```suggestion
            <<<'SQL'
    SELECT DISTINCT m0_.id AS id_0, m0_.title + 0 AS sclr_1
      FROM MyBlogPost m0_
INNER JOIN Category c1_ ON m0_.category_id = c1_.id
  GROUP BY m0_.id
  ORDER BY m0_.title + 0 ASC
SQL
```

File: tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php (hunk 3 above)
Comment (Seb33300) [reply tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php]: I wanted to keep consistency with others queries.
I we want to rewrite queries on multilines, we should do it everywhere.

Comment (greg0ire) [reply tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php]: Not all queries should be multiline, just the long ones.

Comment (Seb33300) [reply tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php]: Yes, but if you look at others existing queries in the same file, they are all long queries.
That's why I think it should be done in a separate PR if you want to rewrite all queries on multiline.

Comment (greg0ire) [reply tests/Doctrine/Tests/ORM/Tools/Pagination/LimitSubqueryWalkerTest.php]: That's an issue I often encounter as a maintainer: we make a style mistake, and then it's repeated over and over for consistency's sake, until someone goes through the painful task of rewriting it all properly. And the longer you wait, the more painful the task is (for instance here 10 queries to rewrite before, 13 after). And often, there is no actual rule in the style guide that says we should write things in some kind of way. In that case, people probably did not know any better. 

Luckily, someone already broke the unworded rule:
https://github.com/doctrine/orm/blob/21bc805dc9662ff05f3038b30a29e05591838a29/tests/Doctrine/Tests/ORM/Functional/LifecycleCallbackTest.php#L190-L199
So don't worry with being inconsistent with other occurrences, the project is already inconsistent.

**Review by lcobucci — CHANGES_REQUESTED** [review]
@Seb33300 thanks for sending this PR but I believe this to be invalid.

As documented in https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/dql-doctrine-query-language.html#doctrine-query-language users are supposed to use `HIDDEN` fields to be able to have arithmetic expression in ORDER clause.

By changing the queries of your test to use this feature the test passes without requiring any modification in the source.

<details>

```php
<?php

declare(strict_types=1);

namespace Doctrine\Tests\ORM\Functional\Ticket;

use Doctrine\ORM\Tools\Pagination\Paginator;
use Doctrine\Tests\Models\CMS\CmsArticle;
use Doctrine\Tests\OrmFunctionalTestCase;
use function count;
use function iterator_to_array;

class GH7805Test extends OrmFunctionalTestCase
{
    protected function setUp()
    {
        $this->useModelSet('cms');
        parent::setUp();
    }

    public function testPaginationWithFunction()
    {
        $article = new CmsArticle();

        $article->topic   = 'Test ORDER BY with Function';
        $article->text    = 'This test fails on MySQL if ORDER BY part is not added to SELECT.';
        $article->version = 1;

        $this->_em->persist($article);
        $this->_em->flush();

        $query = $this->_em->createQuery('SELECT a, SQRT(a.version) AS HIDDEN function_result FROM Doctrine\Tests\Models\CMS\CmsArticle a ORDER BY function_result ASC');
        $query->setFirstResult(0);
        $query->setMaxResults(1);

        $paginator = new Paginator($query, true);
        $paginator->setUseOutputWalkers(false);
        $this->assertEquals(1, count(iterator_to_array($paginator)));
    }

    public function testPaginationWithSimpleArithmetic()
    {
        $article = new CmsArticle();

        $article->topic   = 'Test ORDER BY with SimpleArithmetic';
        $article->text    = 'This test fails on MySQL if ORDER BY part is not added to SELECT.';
        $article->version = 1;

        $this->_em->persist($article);
        $this->_em->flush();

        $query = $this->_em->createQuery('SELECT a, a.version + 0 AS HIDDEN simple_arithmetic FROM Doctrine\Tests\Models\CMS\CmsArticle a ORDER BY simple_arithmetic ASC');
        $query->setFirstResult(0);
        $query->setMaxResults(1);

        $paginator = new Paginator($query, true);
        $paginator->setUseOutputWalkers(false);
        $this->assertEquals(1, count(iterator_to_array($paginator)));
    }
}
```
</details>

**Comment by lcobucci** [comment]
I'll close this as invalid as per my comment. We can discuss things further if you want :+1:

**Pull request closed without merging**